api/src/main/java/jakarta/data/repository/CrudRepository.java

```diff
@@ -38,35 +38,48 @@ public interface CrudRepository<T, K> extends BasicRepository<T, K> {
 
     /**
      * <p>Inserts an entity into the database. If an entity of this type with the same
-     * unique identifier already exists in the database, then this method raises
-     * {@link EntityExistsException}.</p>
+     * unique identifier already exists in the database and the database supports ACID transactions,
+     * then this method raises {@link EntityExistsException}. In databases that follow the BASE model
+     * or use an append model to write data, this exception may not be thrown.</p>
+     *
+     * <p>The entity instance returned as a result of this method may be the same instance as the one
+     * supplied as a parameter, especially in non-Java record classes. However, for Jakarta Data providers
+     * that support Java records, a different instance may be returned.</p>
      *
      * @param entity the entity to insert. Must not be {@code null}.
-     * @throws EntityExistsException if the entity is already present in the database.
+     * @param <S> Type of the entity to insert.
+     * @return the inserted entity, which may or may not be a different instance depending on the Jakarta Data provider.
+     * @throws EntityExistsException if the entity is already present in the database (in ACID-supported databases).
      * @throws NullPointerException if the entity is null.
-     * @throws UnsupportedOperationException for Key-Value and Wide-Column databases
-     *         that use an append model to write data.
      */
-    void insert(T entity);
+    <S extends T> S insert(S entity);
```

Comment:
Please, allow me to give more context for both:

# Insert

We can create an entity, and this database can generate an ID using a strong consistency autoincrement or an ID based on a hash such as UUID.

In NoSQL, this is trivial. 
E.g.: 

```java
record Book(UUID id, String title, String author){} 

var book = Book.of("Effective Java", "Josh");
repository.insert(book);
```

In this case, it will map and convert to the database, and the database will return to the current stage:

```mermaid
    sequenceDiagram
    Entity->>Mapper: The entity from the user
    Mapper->>Database: the conversion
    Database->>Mapper: the database will generate the ID
    Mapper->>Entity: the return will either keep the same instance to non-record or a new instance to record
```

I am not reading but returning to the current stage from the database.

We have a similar case with the version or timestamp at the update method; this may vary the consistency strategy and the algorithm, for example, the clock vector.


I believe that SQL has a similar process.
I mean, they have GeneratedValue and Version annotations.

Do you think it makes sense?

Reply:
> I believe that SQL has a similar process.
> I mean, they have GeneratedValue and Version annotations.

`GeneratedValue` and `Version` annotations are from JPA/Jakarta Persistence, not from SQL, and they are more limited than what SQL can do.  We cannot assume that every Jakarta Data provider will want to limit itself to a JPA-based implementation (and even if they did, the JPA insert operation entitymanager.persist actually returns `void`, not an updated entity!).  SQL allows columns to be generated by the database beyond the Id and Version, and these would need to be retrieved.

In general, I'm fine with having some API in Jakarta Data that is intended for a particular subset of databases and not others.  However, you already have the `save`/`saveAll` operations to cover the pattern that you are trying to achieve for the databases that can optimize this way.  It is not right to force that same pattern upon `insert` and `update` where doing so will prevent this other other subset of databases from being usable or cause them to perform badly.  The advantage of having separate methods for `save` vs `insert` (and `update`) is that we can make the behavior different, and that way cover both subsets of databases with the combination of different behaviors that we offer.

Reply:
> > I believe that SQL has a similar process.
> > I mean, they have GeneratedValue and Version annotations.
> 
> `GeneratedValue` and `Version` annotations are from JPA/Jakarta Persistence, not from SQL, and they are more limited than what SQL can do. We cannot assume that every Jakarta Data provider will want to limit itself to a JPA-based implementation (and even if they did, the JPA insert operation entitymanager.persist actually returns `void`, not an updated entity!). SQL allows columns to be generated by the database beyond the Id and Version, and these would need to be retrieved.
> 
> In general, I'm fine with having some API in Jakarta Data that is intended for a particular subset of databases and not others. However, you already have the `save`/`saveAll` operations to cover the pattern that you are trying to achieve for the databases that can optimize this way. It is not right to force that same pattern upon `insert` and `update` where doing so will prevent this other other subset of databases from being usable or cause them to perform badly. The advantage of having separate methods for `save` vs `insert` (and `update`) is that we can make the behavior different, and that way cover both subsets of databases with the combination of different behaviors that we offer.


When I mentioned the GeneratedValue and Version and mentioned NoSQL databases, for example:
- MongoDB: that provides [ObjectID](https://www.mongodb.com/docs/manual/reference/method/ObjectId/), so when I **insert** the database will provide this for me.
- ArangoDB: when you **update** it will provide a new [Document revision](https://docs.arangodb.com/3.11/concepts/data-structure/documents/#document-revisions)


SQL also provides this behavior with, for example, when I insert with [PostgreSQL with serial](https://www.geeksforgeeks.org/postgresql-serial/), also with [MySQL provider auto increment](https://dev.mysql.com/doc/refman/8.0/en/example-auto-increment.html).


What I don't understand is, why it is specific to NoSQL if SQL has the same approach, could you explain?
The goal is not to support Version and autoincrement. If yes, we need to put this on the insert and update methods.

Reply:
@otaviojava I am not arguing against the usefulness of the one pattern you keep bringing up.  The point is that we already covered that pattern with `save`/`saveAll`, which is inherited from the `BasicRepository`.  We should not be getting rid of the other useful patterns that `insert` and `update` offer and replacing them with a duplication of `save`.  Maybe it will help to list different patterns/goals that a user might have:

| Goal | Method(s) |
| ---- | --------- |
| Create a new entity that has one or more generated values, where the Id is one of the generated values, and continue using the entity afterward | `E save(E)` |
| Create a new entity that has one or more generated values, where the Id is one of the generated values, without using the entity afterward | `void insert(E)` |
| Create a new entity that has no generated values | `void insert(E)` |
| Create a new entity that has one or more generated values, where the Id is not one of the generated values, without using the entity afterward | `void insert(E)` |
| Create a new entity that has one or more generated values, where the Id is not one of the generated values, and continue using the entity afterward | `void insert(E)`, then `E findById(id)` |
| Create a new entity that has a specified Id or otherwise update the existing entity | `E save(E)` |
| If an entity (with generated values) with the specified id exists, modify the entity and continue using the entity afterward | boolean `update(E)`, then `E findById(id)` |
| If an entity (without generated values) with the specified id exists, modify the entity and continue using the entity afterward | boolean `update(E)` |
| If an entity with the specified id exists, modify the entity and do not use the entity afterward | boolean `update(E)` |

The different signatures for `E save(E)`, `void insert(E)`, and `boolean update(E)` allow the user to achieve every one of the scenarios above and lead the user to naturally select which is most optimal to their pattern.  The proposal to make insert return the entity will cause the 4th, 8th, and 9th above to become inefficient wherever the database or programming model doesn't have a special optimization to return **all** of the generated values.  It will require additional retrievals from the database, and wherever this happens, users will despise it.  If you don't believe it will be very common, I've already pointed out JPA's only pure insert operation, [EntityManager.persist](https://jakarta.ee/specifications/persistence/3.1/apidocs/jakarta.persistence/jakarta/persistence/entitymanager#persist(java.lang.Object)), returns a `void` type, not an updated entity.  Furthermore, JPA doesn't not have this optimization for pure [updates](https://jakarta.ee/specifications/persistence/3.1/apidocs/jakarta.persistence/jakarta/persistence/query#executeUpdate()) either.  A JDBC-based implementation might or might not have an optimization available, depending on the capabilities of the database and JDBC driver.  It is wrong to make core API force inefficient behavior here which users will be very displeased with.  I don't agree with this change.  If there needs to be any further discussion of it, let's schedule a call to go over it because we keep restating the same things over and over in the comments.

Reply:
The EntityManager.persist will return it because it changes the class that must be mutable as the spec says:

> An update to the state of an entity includes both the assignment of a new value to a persistent property or field of the entity as well as the modification of a mutable value of a persistent property or field

https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#a1955

In that case, it will be easier to implement JPA IMHO:

```java
@Overwrite
public Movie insert(Movie movie){
//no id, because it will be auto-generated by the database
entityManager.insert(movie); that changes will happen in this class because it is mutable
return movie;
}
```
Thus, the generated ID and version will happen in this instance. Did I get it right?

Imagine a scenario in which I want to **insert** it, but with a mutable class using record.

```java
@Overwrite
public Movie insert(Movie movie){
//no id, because it will be auto-generated by the database
otherDatabaseprovider.insert(movie);// I must create a new instance because this is a record.
return movieUpdated;
}
```

If I don't have this return, I will lose track, mainly if this is a hashed ID.

If we work with JPA, I agree that it can directly change the entity itself once it does not support records, but if any provider does, this insert will become impossible to use. 

If you take a look at the Micronaut, you can see that it will return the entity:

https://micronaut-projects.github.io/micronaut-data/2.3.0/api/index.html?io/micronaut/data/repository/CrudRepository.html

Reply:
> The EntityManager.persist will return void, because it changes the class that **must** be mutable as the spec says:
> 
> > An update to the state of an entity includes both the assignment of a new value to a persistent property or field of the entity as well as the modification of a mutable value of a persistent property or field
> 
> https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#a1955

I understood this to be saying that you can update entities by assigning a new value to a field/property,

```
movie.title = "Updated Title";
```

or modify the value of a mutable field or property,

```
movie.actors.add("Someone Famous");
```

But if your interpretation is correct, (it appears that it might be based on trying it out on one JPA provider which does modify the supplied entity instance) then we would be able to allow `E insert(E)`, but I still don't see a way to allow the entity return type on `update`.